pallets/dapp-staking/src/lib.rs

```diff
@@ -1516,6 +1533,157 @@ pub mod pallet {
+                amount_to_move,
+                current_era,
+                protocol_state.subperiod(),
+            );
+
+            ensure!(
+                dest_staking_info.total_staked_amount() >= T::MinimumStakeAmount::get(),
+                Error::<T>::InsufficientStakeAmount
+            );
+
+            if is_new_entry && !is_source_unregistered {
+                ledger.contract_stake_count.saturating_inc();
+                ensure!(
+                    ledger.contract_stake_count <= T::MaxNumberOfStakedContracts::get(),
+                    Error::<T>::TooManyStakedContracts
+                );
+            }
+
+            // 3.
+            // Handle Bonus Status
+            // For an unregistered contract the bonus status is preserved.
+            // For a registered contract, the source unstake has already handled the bonus status logic.
+            dest_staking_info.bonus_status = if is_source_unregistered {
+                bonus_status_snapshot
+            } else {
+                source_staking_info.bonus_status.clone()
+            };
+
+            // 4.
+            // Update Afected Contract Stakes
+            if let Some(source_dapp_info) = maybe_source_dapp_info {
+                // Registered source: perform unstake operations.
+                let mut source_contract_stake_info = ContractStake::<T>::get(source_dapp_info.id);
+                source_contract_stake_info.unstake(
+                    era_and_amount_pairs,
+                    protocol_state.period_info,
+                    current_era,
+                );
+
+                ContractStake::<T>::insert(&source_dapp_info.id, source_contract_stake_info);
+            }
+
+            let mut dest_contract_stake_info = ContractStake::<T>::get(&dest_dapp_info.id);
+            dest_contract_stake_info.stake(amount_to_move, protocol_state.period_info, current_era);
+
+            // 5.
+            // Update remaining storage entries
+            if !is_source_unregistered && source_staking_info.is_empty() {
+                ledger.contract_stake_count.saturating_dec();
+                StakerInfo::<T>::remove(&account, &source_contract);
+            } else if !is_source_unregistered {
+                StakerInfo::<T>::insert(&account, &source_contract, source_staking_info);
+            } else {
+                // Unregistered source: remove staker info directly
+                StakerInfo::<T>::remove(&account, &source_contract);
+            }
+
+            StakerInfo::<T>::insert(&account, &destination_contract, dest_staking_info);
+
+            Self::update_ledger(&account, ledger)?;
+            ContractStake::<T>::insert(&dest_dapp_info.id, dest_contract_stake_info);
+
+            Self::deposit_event(Event::<T>::StakeMoved {
+                account,
+                source_contract,
+                destination_contract,
+                amount: amount_to_move,
+            });
+
+            Ok(())
+        }
```

Comment:
General comment about the extrinsic call - we should reuse code more.
This is essentially `unstake` and `stake`.

Logic should be extracted from those calls, and reused here.
Something like this: https://github.com/AstarNetwork/astar-frame/blob/polkadot-v0.9.39/frame/dapps-staking/src/pallet/mod.rs#L652